pkg/controller/workload/mpijob/mpijob_controller.go

```diff
@@ -0,0 +1,579 @@
+		return err
+	}
+
+	r.record.Eventf(job, corev1.EventTypeNormal, "CreatedWorkload",
+		"Created Workload: %v", workload.Key(wl))
+	return nil
+}
+
+// ensureAtMostOneWorkload finds a matching workload and deletes redundant ones.
+func (r *MPIJobReconciler) ensureAtMostOneWorkload(ctx context.Context, job *kubeflow.MPIJob) (*kueue.Workload, error) {
+	log := ctrl.LoggerFrom(ctx)
+
+	// Find a matching workload first if there is one.
+	var toDelete []*kueue.Workload
+	var match *kueue.Workload
+
+	var workloads kueue.WorkloadList
+	if err := r.client.List(ctx, &workloads, client.InNamespace(job.Namespace),
+		client.MatchingFields{ownerKey: job.Name}); err != nil {
+		log.Error(err, "Unable to list child workloads")
+		return nil, err
+	}
+
+	for i := range workloads.Items {
+		w := &workloads.Items[i]
+		owner := metav1.GetControllerOf(w)
+		// Indexes don't work in unit tests, so we explicitly check for the
+		// owner here.
+		if owner.Name != job.Name {
+			continue
+		}
+		if match == nil && jobAndWorkloadEqual(job, w) {
+			match = w
+		} else {
+			toDelete = append(toDelete, w)
+		}
+	}
+
+	// If there is no matching workload and the job is running, suspend it.
+	if match == nil && !jobSuspended(job) {
+		log.V(2).Info("job with no matching workload, suspending")
+		var w *kueue.Workload
+		if len(workloads.Items) == 1 {
+			// The job may have been modified and hence the existing workload
+			// doesn't match the job anymore. All bets are off if there are more
+			// than one workload...
+			w = &workloads.Items[0]
+		}
+		if err := r.stopJob(ctx, w, job, "No matching Workload"); err != nil {
+			log.Error(err, "stopping job")
+		}
+	}
+
+	// Delete duplicate workload instances.
+	existedWls := 0
+	for i := range toDelete {
+		err := r.client.Delete(ctx, toDelete[i])
+		if err == nil || !apierrors.IsNotFound(err) {
+			existedWls++
+		}
+		if err != nil && !apierrors.IsNotFound(err) {
+			log.Error(err, "Failed to delete workload")
+		}
+		if err == nil {
+			r.record.Eventf(job, corev1.EventTypeNormal, "DeletedWorkload",
+				"Deleted not matching Workload: %v", workload.Key(toDelete[i]))
+		}
+	}
+
+	if existedWls != 0 {
+		if match == nil {
+			return nil, fmt.Errorf("no matching workload was found, tried deleting %d existing workload(s)", existedWls)
+		}
+		return nil, fmt.Errorf("only one workload should exist, found %d", len(workloads.Items))
+	}
+
+	return match, nil
+}
+
+func ConstructWorkloadFor(ctx context.Context, client client.Client,
+	job *kubeflow.MPIJob, scheme *runtime.Scheme) (*kueue.Workload, error) {
+	w := &kueue.Workload{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      GetWorkloadNameForMPIJob(job.Name),
+			Namespace: job.Namespace,
+		},
+		Spec: kueue.WorkloadSpec{
+			QueueName: queueName(job),
+		},
+	}
+
+	highestPriorityClassName := ""
+	highestPriority := int32(0)
+
+	for index, mpiReplicaType := range orderedReplicaTypes(&job.Spec) {
+		podSet := kueue.PodSet{
+			Name:  strconv.Itoa(index),
```

Comment:
```suggestion
			Name:  mpiReplicaType,
```

This will make it easier to visually match

Reply:
I was thinking about it, but `mpiReplicaType` is actually a free string, which does not need to be a valid CRD name. Even in this case we have `Launcher` which isn't valid due to the capital letter. I could lower case it, but lowecasing wouldn't work in general.

Reply:
sigh... in any case mpi operator only cares about two strings: Launcher and Worker, so we can hardcode them.

Also keep in mind that we plan to replace the map with proper fields in the future https://github.com/kubeflow/mpi-operator/issues/519

Reply:
ok, lowercased to make this work.